File: Src/Engine/Editor/SceneGraph/ActorGraphNode.cpp

#include "ActorGraphNode.h"

#include "ActorChildNode.h"
#include "Editor/GUI/Tree/ActorTreeNode.h"
#include "Editor/Modules/SceneGraphModule.h"
#include "Runtime/Level/Actor.h"
#include "Runtime/Level/Scene/Scene.h"
#include "runtime/ui/gui/rootcontrol.h"

namespace SE::Editor
{
	String ActorGraphNode::__GetName()
	{
		return _actor->GetName();;
	}

	ScenesGraphNode* ActorGraphNode::__GetParentScene()
	{
		Scene* scene = _actor ? _actor->GetScene() : nullptr;
		return scene != nullptr ? TypeCast<ScenesGraphNode>(EditorApp::Ins().sceneGraphModule->FindNode(scene->GetInstanceID())) : nullptr;
	}

	ActorGraphNode* ActorGraphNode::Find(::SE::Actor* actor)
	{
		// Check itself
		if (_actor == actor)
			return this;

		// Check deeper
		for (int i = 0; i < ChildNodes.Count(); i++)
		{
			ActorGraphNode* node;
			if (TypeTryCast(ChildNodes[i], node))
			{
				ActorGraphNode* result = node->Find(actor);
				if (result != nullptr)
				{
					return result;
				}
			}
		}

		return nullptr;
	}

	ActorChildNode* ActorGraphNode::AddChildNode(ActorChildNode* node)
	{
		ActorChildNodes.Add(node);
		node->ParentNode = this;
		return node;
	}

	void ActorGraphNode::DisposeChildNodes()
	{
		// Send event to root so if any of this child nodes is selected we can handle it
		RootControl* root = nullptr ;//Root;
		if (root != nullptr)
		{
			// root->OnActorChildNodesDispose(this);
		}

		if (!ActorChildNodes.IsEmpty())
		{
			for (int i = 0; i < ActorChildNodes.Count(); i++)
			{
				ActorChildNode* node = ActorChildNodes[i];
				node->Dispose();
				Delete(node);
			}
			ActorChildNodes.Clear();
		}
	}

	ActorGraphNode* ActorGraphNode::FindChildActor(::SE::Actor* actor)
	{
		for (int i = 0; i < ChildNodes.Count(); i++)
		{
			ActorGraphNode* node;
			if (TypeTryCast(ChildNodes[i], node) && node->Actor == actor)
			{
				return node;
			}
		}

		return nullptr;
	}
	Object* ActorGraphNode::GetEditableObject()
	{
		return _actor;
	}

	void ActorGraphNode::LinkTreeNode()
	{
		_treeNode->LinkNode(this);
	}

	ActorGraphNode::ActorGraphNode() : ScenesGraphNode(UID::New())
	{

	}

	ActorGraphNode::ActorGraphNode(::SE::Actor* actor) : ScenesGraphNode(actor->GetInstanceID())
	{
		_actor = actor;
		_treeNode = New<ActorTreeNode>();
	}

	ActorGraphNode::ActorGraphNode(::SE::Actor* actor, UID id) : ScenesGraphNode(id)
	{
		_actor = actor;
		_treeNode = New<ActorTreeNode>();
	}

	ActorGraphNode::ActorGraphNode(::SE::Actor* actor, ActorTreeNode* treeNode) : ScenesGraphNode(actor->GetInstanceID())
	{
		_actor = actor;
		_treeNode = treeNode;
	}

	void ActorGraphNode::OnParentChanged()
	{
		_treeNode->OnParentChanged(_actor, TypeCast<ActorGraphNode>(parentNode));
	}
} // SE

File: Src/Engine/Editor/SceneGraph/ActorGraphNode.cs
using System;
using System.Collections.Generic;
using SE.Editor.GUI;

namespace SE.Editor.SceneGraph
{
    /// <summary>
    /// Scene graph node backed by a native Runtime actor.
    /// </summary>
    public class ActorGraphNode : ScenesGraphNode
    {
        private SE.Actor? m_Actor;
        private readonly ActorTreeNode m_TreeNode;
        private bool m_IsTreeNodeLinked;

        public ActorGraphNode(SceneGraphFactory factory, SE.Actor actor)
            : this(factory, actor, actor?.SceneObjectId ?? throw new ArgumentNullException(nameof(actor)), new ActorTreeNode())
        {
        }

        internal ActorGraphNode(SceneGraphFactory factory, SE.Actor? actor, Guid id)
            : this(factory, actor, id, new ActorTreeNode())
        {
        }

        protected ActorGraphNode(SceneGraphFactory factory, SE.Actor? actor, Guid id, ActorTreeNode treeNode)
            : base(factory, id)
        {
            m_Actor = actor;
            m_TreeNode = treeNode ?? throw new ArgumentNullException(nameof(treeNode));
        }

        public SE.Actor? Actor => m_Actor;
        public ActorTreeNode TreeNode => m_TreeNode;
        public List<ActorChildNode> ActorChildNodes { get; } = new();

        public override string Name => m_Actor?.Name ?? string.Empty;

        public override SceneGraphNode? ParentScene
        {
            get
            {
                SE.Scene? scene = m_Actor?.Scene;
                return scene != null ? Factory.FindNode(scene.SceneObjectId) as SceneGraphNode : null;
            }
        }

        public ActorGraphNode? Find(SE.Actor actor)
        {
            ArgumentNullException.ThrowIfNull(actor);
            if (ReferenceEquals(m_Actor, actor) || m_Actor?.SceneObjectId == actor.SceneObjectId)
                return this;

            foreach (ScenesGraphNode child in ChildNodes)
            {
                if (child is ActorGraphNode actorNode)
                {
                    ActorGraphNode? result = actorNode.Find(actor);
                    if (result != null)
                        return result;
                }
            }

            return null;
        }

        public ActorChildNode AddChildNode(ActorChildNode node)
        {
            ArgumentNullException.ThrowIfNull(node);
            ActorChildNodes.Add(node);
            node.ParentNode = this;
            return node;
        }

        public void DisposeChildNodes()
        {
            ActorChildNode[] childNodes = ActorChildNodes.ToArray();
            foreach (ActorChildNode child in childNodes)
                child.Dispose();
            ActorChildNodes.Clear();
        }

        public ActorGraphNode? FindChildActor(SE.Actor actor)
        {
            ArgumentNullException.ThrowIfNull(actor);
            foreach (ScenesGraphNode child in ChildNodes)
            {
                if (child is ActorGraphNode node &&
                    (ReferenceEquals(node.Actor, actor) || node.Actor?.SceneObjectId == actor.SceneObjectId))
                {
                    return node;
                }
            }

            return null;
        }

        public override SE.Object? GetEditableObject()
        {
            return m_Actor;
        }

        public void LinkTreeNode()
        {
            if (m_IsTreeNodeLinked)
                return;

            m_IsTreeNodeLinked = true;
            m_TreeNode.LinkNode(this);
        }

        internal void RefreshActor(SE.Actor actor)
        {
            m_Actor = actor ?? throw new ArgumentNullException(nameof(actor));
            m_TreeNode.UpdateText();
        }

        protected override void OnParentChanged()
        {
            m_TreeNode.OnParentChanged(ParentNode as ActorGraphNode);
        }

        public override void Dispose()
        {
            if (IsDisposed)
                return;

            m_TreeNode.Parent = null;

            m_TreeNode.Dispose();
            DisposeChildNodes();
            base.Dispose();
        }

        public override void OnDispose()
        {
            if (IsDisposed)
                return;

            ActorChildNodes.Clear();
            m_Actor = null;
            base.OnDispose();
        }
    }
}
